pyteomics/proforma.py

```diff
@@ -0,0 +1,1501 @@
+'''
+proforma - Proteoform and Peptidoform Notation
+==============================================
+
+ProForma is a notation for defining modified amino acid sequences using
+a set of controlled vocabularies, as well as encoding uncertain or partial
+information about localization. See `ProForma specification <https://www.psidev.info/proforma>`_
+for more up-to-date information.
+
+Strictly speaking, this implementation supports ProForma v2.
+
+Data Access
+-----------
+
+:py:func:`parse_proforma` - The primary interface for parsing ProForma strings.
+
+    >>> parse_proforma("EM[Oxidation]EVT[#g1(0.01)]S[#g1(0.09)]ES[Phospho#g1(0.90)]PEK")
+        ([('E', None),
+          ('M', GenericModification('Oxidation', None, None)),
+          ('E', None),
+          ('V', None),
+          ('T', LocalizationMarker(0.01, None, '#g1')),
+          ('S', LocalizationMarker(0.09, None, '#g1')),
+          ('E', None),
+          ('S',
+          GenericModification('Phospho', [LocalizationMarker(0.9, None, '#g1')], '#g1')),
+          ('P', None),
+          ('E', None),
+          ('K', None)],
+         {'n_term': None,
+          'c_term': None,
+          'unlocalized_modifications': [],
+          'labile_modifications': [],
+          'fixed_modifications': [],
+          'intervals': [],
+          'isotopes': [],
+          'group_ids': ['#g1']})
+
+:py:func:`to_proforma` - Format a sequence and set of properties as ProForma text.
+
+
+Classes
+-------
+
+:py:class:`ProForma` - An object oriented version of the parsing and formatting code,
+coupled with minimal information about mass and position data.
+'''
+
+from pyteomics.mass.mass import calculate_mass
+import re
+import warnings
+from collections import namedtuple, defaultdict, deque
+from functools import partial
+
+try:
+    from enum import Enum
+except ImportError:
+    # Python 2 doesn't have a builtin Enum type
+    Enum = object
+
+
+from pyteomics import parser
+from pyteomics.mass import Composition, std_aa_mass, Unimod
+from pyteomics.auxiliary import PyteomicsError, BasicComposition
+from pyteomics.auxiliary.utils import add_metaclass
+
+# To eventually be implemented with pyteomics port?
+try:
+    from psims.controlled_vocabulary.controlled_vocabulary import (load_psimod, load_xlmod, load_gno, obo_cache)
+except ImportError:
+    def _needs_psims(name):
+        raise ImportError("Loading %s requires the `psims` library. To access it, please install `psims" % name)
+
+    load_psimod = partial(_needs_psims, 'PSIMOD')
+    load_xlmod = partial(_needs_psims, 'XLMOD')
+    load_gno = partial(_needs_psims, 'GNO')
+    obo_cache = None
```

Comment:
This name does not seem to be used, is it necessary?

Reply:
The name gets baked into the `partial`-made function so that the error is clear about which "entity" depended upon the other source, e.g.
```python 
>>> load_psimod()
ImportError: Loading PSIMOD requires the `psims` library. To access it, please install `psims`
```
Technically, we could just make the message "Loading this controlled vocabulary requires `psims`." and be done, but it feels less explicit.

Reply:
To be clear, this comment of mine referred to `obo_cache` only. It is not used in `pyteomics` code. Not sure why Github displayed four lines like that.

Reply:
Thank you for the clarifying. Yes, that variable is imported to allow `proforma` to expose control over the default file cache from `psims`. That way users could set the cache directory or disable the cache altogether if they wished. Otherwise they'd need to explicitly import it from `psims`. It probably isn't necessary to import it here and just needs to be documented clearly that it interacts with the `psims` cache mechanism.

Reply:
Thank you! Indeed, I think it's worth mentioning in the docs because the user may not even know that `psims` is used, or how it works with caches.